src/ndisc/nm-ndisc.c

```diff
@@ -421,20 +422,33 @@ nm_ndisc_add_address (NMNDisc *ndisc, const NMNDiscAddress *new)
 	for (i = 0; i < rdata->addresses->len; i++) {
 		NMNDiscAddress *item = &g_array_index (rdata->addresses, NMNDiscAddress, i);
 
-		if (IN6_ARE_ADDR_EQUAL (&item->address, &new->address)) {
-			if (new->lifetime == 0) {
-				g_array_remove_index (rdata->addresses, i);
-				return TRUE;
-			}
-
-			if (   item->dad_counter == new->dad_counter
-			    && get_expiry (item) == get_expiry (new)
-			    && get_expiry_preferred (item) == get_expiry_preferred (new))
-				return FALSE;
+		if (from_ra) {
+			/* RFC4862 5.5.3.d, we find an existing address with the same prefix.
+			 * (note that all prefixes at this point have implicity length /64). */
+			if (memcmp (&item->address, &new->address, 8) != 0)
+				continue;
+		} else {
+			if (!IN6_ARE_ADDR_EQUAL (&item->address, &new->address))
+				continue;
+		}
 
-			*item = *new;
+		if (new->lifetime == 0) {
+			g_array_remove_index (rdata->addresses, i);
 			return TRUE;
 		}
+
+		if (   get_expiry (item) == get_expiry (new)
+		    && get_expiry_preferred (item) == get_expiry_preferred (new)
+		    && (   from_ra
+		        || item->dad_counter != new->dad_counter))
```

Comment:
Shouldn't this be `item->dad_counter == new->dad_counter` ?

Reply:
true